cli/cluster/cluster.go

```diff
@@ -0,0 +1,145 @@
+								if instance == nil {
+									instance = NewConfig()
+								}
+								instance.Set(path[1:], value)
+							}
+						} else if rname == path[0] {
+							replicaset.Set(path[1:], value)
+						}
+					}
+				} else if gname == path[0] {
+					group.Set(path[1:], value)
+				}
+			}
+		} else {
+			global.Set(path, value)
+		}
+	})
+
+	if instance == nil {
+		return nil
+	}
+
+	instance.Merge(replicaset)
+	instance.Merge(group)
+	instance.Merge(global)
+
+	return instance
+}
+
+// HasInstance returns true if an instance with the name exists in the config.
+func HasInstance(config *Config, name string) bool {
+	return findInstance(config, name) != nil
+}
+
+// Instantiate returns a fetched instance config from the cluster config. If
+// the cluster config has the instance then it returns a merged config of the
+// instance from scopes: global, group, replicaset, instance. If the cluster
+// config has not the instance then it returns a global scope of the cluster
+// config.
+func Instantiate(config *Config, name string) *Config {
+	iconfig := findInstance(config, name)
+	if iconfig != nil {
+		return iconfig
+	}
+
+	iconfig = NewConfig()
+	config.ForEach(nil, func(path []string, value any) {
+		if len(path) < 1 || path[0] != "groups" {
+			iconfig.Set(path, value)
+		}
+	})
+
+	return iconfig
+}
+
+// GetClusterConfig returns a cluster configuration loaded from a path to
+// a config file. It uses a a config file, etcd and default environment
+// variables as sources. The function returns a cluster config as is, without
+// merging of settings from scopes: global, group, replicaset, instance.
+func GetClusterConfig(path string) (*Config, error) {
+	if path == "" {
+		return nil, fmt.Errorf("a configuration file must be set")
+	}
+
+	collector := NewFileCollector(path)
+	config, err := collector.Collect()
+	if err != nil {
+		fmtErr := "unable to get cluster config from %q: %s"
+		return nil, fmt.Errorf(fmtErr, path, err)
+	}
+
+	defaultEnvCollector := NewEnvCollector(func(path []string) string {
+		middle := strings.ToUpper(strings.Join(path, "_"))
+		return fmt.Sprintf("TT_%s_DEFAULT", middle)
+	})
+	defaultEnvConfig, err := defaultEnvCollector.Collect()
+	if err != nil {
+		fmtErr := "failed to collect a config from default environment variables: %w"
+		return nil, fmt.Errorf(fmtErr, err)
+	}
+
+	config.Merge(defaultEnvConfig)
+	return config, err
+}
+
+// GetInstanceConfig returns a full configuration for an instance with the
+// name from a cluster config. It merges the configuration from all configured
+// sources and scopes: environment, global, group, replicaset, instance.
+func GetInstanceConfig(clusterConfig *Config, instance string) (*Config, error) {
+	if !HasInstance(clusterConfig, instance) {
+		return nil, fmt.Errorf("an instance %q not found", instance)
+	}
+
+	mainEnvCollector := NewEnvCollector(func(path []string) string {
+		middle := strings.ToUpper(strings.Join(path, "_"))
+		return fmt.Sprintf("TT_%s", middle)
```

Comment:
Will this merge `TT_CONFIG_*` vars to my instance config? If yes, this is incorrect. `config` section is only a global level section. Yes, I can just ignore this section for instance level, but if I ever want to dump this instance to YAML, it will be incorrect config.

Reply:
The same logic in the implementation of Tarantool's cluster/instance config.

There is some logic in this too: a fully complete configuration of an instance is impossible without `config` section. Yes, this section could only be described at the global scope (like `sharding` only on global + group). But `config` like `sharding` is a part of the full configuration of the instance. See:

https://www.notion.so/tarantool/config-v3-0-0-cb0a2e8f1cf94cdf9e19416c5f77618d

It's not difficult for me to make changes in the implementation, up to you.

But it seems to me it would be more logical to nil `config` and `sharding` sections before export to yaml as an instance scope configuration.

Reply:
> config and sharding sections before export to yaml as an instance scope configuration.

And 'config' in replicasets and in groups, and 'sharding' in replicasets. Shouldn't this be controlled by schema, not manual setting to nil?

Reply:
See:

https://github.com/tarantool/tarantool/blob/671db638b8bb155b3329b11a47a90ac44d47138a/src/box/lua/config/instance_config.lua#L246-L250

Reply:
This will be restricted as I know.